app/jobs/autoscaling.py

```diff
@@ -260,227 +61,313 @@ def get_instance_tags():
+    if not (gsnap_ASG.can_scale and rapsearch_ASG.can_scale):
+        print "Scaling by agents of {my_environment} is not permitted.".format(my_environment=my_environment)
+        return
 
-def stop_draining(asg, tag_list, draining_tag, num_instances):
-    instance_ids = instances_to_rescue(asg, tag_list, draining_tag, num_instances)
-    if instance_ids:
-        print "Stopping drainage of the following instances: " + ",".join(instance_ids)
-        remove_draining_tag(instance_ids, draining_tag)
+    print "{my_num_jobs} jobs are in progress.".format(my_num_jobs=my_num_jobs)
+    compute_num_instances = required_capacity(my_num_jobs)
+    for service_ASG in [gsnap_ASG, rapsearch_ASG]:
+        print "--- Analyzing the {instance_name} ASG ---".format(instance_name=service_ASG.instance_name)
+        num_desired = service_ASG.num_desired_instances()
+        healthy_instances, draining_instances, terminating_instances, corrupt_instances = service_ASG.classify_instances()
+        num_healthy = len(healthy_instances)
+        num_draining = len(draining_instances)
+        print "CURRENTLY:"
+        print "The desired capacity is {num_desired}.".format(num_desired=num_desired)
+        print "There are {num_healthy} healthy instances: {healthy_instances}.".format(num_healthy=num_healthy, healthy_instances=healthy_instances)
+        print "There are {num_draining} draining instances: {draining_instances}.".format(num_draining=num_draining, draining_instances=draining_instances)
+        print "There are {num_terminating} terminating instances: {terminating_instances}.".format(num_terminating=len(terminating_instances), terminating_instances=terminating_instances)
+        assert not corrupt_instances, "There are {num_corrupt} corrupt instances: {corrupt_instances}.".format(num_corrupt=len(corrupt_instances), corrupt_instances=corrupt_instances)
+
+        print "MOVING FORWARD:"
+        raw_new_num_desired = compute_num_instances(num_desired)
+        print "In principle, the desired capacity needs to be {raw_new_num_desired}.".format(raw_new_num_desired=raw_new_num_desired)
+        new_num_desired = service_ASG.clamp_to_valid_range(raw_new_num_desired)
+        print "Enforcing the MinSize and MaxSize set by the operator, the desired capacity will be set to {new_num_desired}.".format(new_num_desired=new_num_desired)
+        service_ASG.set_desired_capacity(new_num_desired)
+
+        print "HEALTHY <--> DRAINING transitions:"
+        service_ASG.draining_instances = draining_instances
+        if new_num_desired < num_healthy:
+            num_to_drain = num_healthy - new_num_desired
+            instances_to_drain = service_ASG.start_draining(num_to_drain)
+            print "{num_to_drain} instances need to be drained: {instances_to_drain}.".format(num_to_drain=num_to_drain, instances_to_drain=instances_to_drain)
+        elif new_num_desired > num_healthy:
```

Comment:
It seems from the `else` print below, you need conjunct `... and num_draining > 0` here.

Reply:
Ok you're right. It is actually required for the logic. Even though the effects are only cosmetic. Changed